### PR TITLE
add on-demand mode (LP: #1824282)

### PR DESCRIPTION
**Pull request by ycheng**
add on-demand mode so that we have more flexibility in using GPU resources.

Signed-off-by: Yuan-Chen Cheng <yc.cheng@canonical.com>

### CODE REVIEW

File: prime-select [hunk 1]
```diff
@@ -126,13 +135,13 @@ class Switcher(object):
 
         return True
 
-    def _disable_nvidia(self):
+    def _disable_nvidia(self, keep_cuda=False):
```
Comment (tseliot) [inline prime-select]: Maybe "keep_cuda" could become "keep_nvidia_modules"?

```suggestion
    def _disable_nvidia(self, keep_nvidia_modules=False):
```

File: prime-select [hunk 2]
```diff
@@ -126,13 +135,13 @@ class Switcher(object):
...
         try:
             os.unlink(self._old_blacklist_file)
         except:
             pass
 
-        self._blacklist_nvidia()
+        self._blacklist_nvidia(keep_cuda)
```
Comment (tseliot) [inline prime-select]:
```suggestion
        self._blacklist_nvidia(keep_nvidia_modules)
```

File: prime-select [hunk 3]
```diff
@@ -150,8 +159,17 @@ class Switcher(object):
         if not os.path.isfile(self._nvidia_kms_file):
             self._enable_kms()
 
-    def _blacklist_nvidia(self):
-        blacklist_text = '''# Do not modify
+    def _blacklist_nvidia(self, keep_cuda=False):
```
Comment (tseliot) [inline prime-select]:
```suggestion
    def _blacklist_nvidia(self, keep_nvidia_modules=False):
```

File: prime-select [hunk 4]
```diff
@@ -114,6 +121,8 @@ class Switcher(object):
             # Always allow enabling nvidia
             # (No need to check if nvidia is available)
             self._enable_nvidia()
+        elif profile == "on-demand":
+            self._disable_nvidia(keep_cuda=True)
```
Comment (tseliot) [inline prime-select]:
```suggestion
            self._disable_nvidia(keep_nvidia_modules=True)
```

File: prime-select (hunk 1 above)
Comment (ycheng) [reply prime-select]: got it , will refine it.

**Comment by ycheng** [comment]
fix accordingly. please check the newly pushed branch.

File: prime-select [hunk 5]
```diff
@@ -150,8 +159,17 @@ class Switcher(object):
         if not os.path.isfile(self._nvidia_kms_file):
             self._enable_kms()
 
-    def _blacklist_nvidia(self):
-        blacklist_text = '''# Do not modify
+    def _blacklist_nvidia(self, keep_nvidia_modules=False):
+        blacklist_text = ""
+        if keep_nvidia_modules:
```
Comment (tseliot) [inline prime-select]: In this case, we should not blacklist any of the nvidia modules. I will make sure that gpu-manager removes the additional output class that causes Nvidia to be used with X in this case.

Comment (ycheng) [reply prime-select]: glad to know that there will be a new complete solution landing in the gpu manager.

fix accordingly. please check the newly pushed branch.

**Comment by tseliot** [comment]
Merged. Thanks!

**Pull request merged**